gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpHeadersUtils.java

```diff
@@ -0,0 +1,103 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.httpjson;
+
+import com.google.api.client.http.HttpHeaders;
+import com.google.api.client.util.ClassInfo;
+import com.google.api.client.util.FieldInfo;
+import com.google.api.core.InternalApi;
+import com.google.common.collect.ImmutableList;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class is for internal use only and is public for technical reasons.
+ */
+@InternalApi
+public class HttpHeadersUtils {
+  public static String getUserAgentValue(Map<String, String> headersMap) {
+    for (Map.Entry<String, String> entry : headersMap.entrySet()) {
+      if ("user-agent".equals(entry.getKey().toLowerCase())) {
+        return entry.getValue();
+      }
+    }
+    return null;
+  }
+
+  public static HttpHeaders setHeaders(HttpHeaders headers, Map<String, String> headersMap) {
+    for (Map.Entry<String, String> entry : headersMap.entrySet()) {
+      setHeader(headers, entry.getKey(), entry.getValue());
+    }
+    return headers;
+  }
+
+  public static HttpHeaders setHeader(HttpHeaders headers, String key, String value) {
+    ClassInfo classInfo = headers.getClassInfo();
+    FieldInfo fieldInfo = classInfo.getFieldInfo(key);
+    Object headerValue;
+    if (fieldInfo == null) {
+      headerValue = value;
+    } else {
+      if (List.class.isAssignableFrom(fieldInfo.getType())) {
+        Type genericType = fieldInfo.getGenericType();
+        if (genericType instanceof ParameterizedType) {
+          ParameterizedType parameterizedType = (ParameterizedType) genericType;
+          // parameterizedType is a list, so is guaranteed to have at least one type argument
+          Type listElementType = parameterizedType.getActualTypeArguments()[0];
+          if (listElementType instanceof Class) {
+            headerValue = ImmutableList.of(fromString((Class) listElementType, value));
+          } else {
+            throw new IllegalArgumentException();
+          }
+        } else {
+          headerValue = ImmutableList.<Object>of(value);
+        }
+      } else {
+        headerValue = fromString(fieldInfo.getType(), value);
+      }
+    }
+    headers.set(key, headerValue);
+    return headers;
+  }
+
+  private static Object fromString(Class<?> clazz, String stringValue) {
+    if (String.class == clazz) {
+      return stringValue;
+    } else if (Long.class == clazz) {
+      return Long.valueOf(stringValue);
+    }
+
+    // It seems only String and Long have chances of being used. Not adding other
+    // classes to avoid having dead code, even though without them this looks incomplete.
+    throw new IllegalArgumentException();
```

Comment:
Can you give more details in this exception?

Reply:
Added description